File: practice/16_3Sum_Closest.js
function closest(nums, target) {
    nums.sort((a, b) => a - b);
    let min = Number.MAX_SAFE_INTEGER,
        result = min;
    for (let i = 0; i < nums.length; i++) {
        let start = i + 1,
            end = nums.length - 1;

        while (start < end) {
            let sum = nums[i] + nums[start] + nums[end];
            let abs = Math.abs(target - sum);
            if (abs < min) {
                min = Math.min(min, abs);
                result = sum;
            }
            if (sum < target) {
                start++;
            } else {
                end--;
            }
        }
    }

    return result;
}

// console.log(closest([-1, 2, 1, -4], 1));
console.log(closest([0, 0, 0], 1));

File: practice/16_3Sum_Closest.kt
fun main() {
	println(closest(intArrayOf(-1, 2, 1, -4), 1))
}

fun closest(nums: IntArray, target: Int): Int {
	nums.sort()
	var result = 0
	var min = Int.MAX_VALUE

	for(i in 0 until nums.size) {
		var start=i+1
		var end=nums.size-1

		while(start<end) {
			val sum = nums[i] + nums[start] + nums[end]
			val diff = Math.abs(target-sum)

			if(diff < min) {
				min = minOf(min, diff)
				result = sum
			}

			when {
				sum < target -> start ++
				sum > target -> end--
				else return sum
			}
		}
	}

	return result
}
